prow/cmd/generic-autobumper/main.go

```diff
@@ -19,47 +19,467 @@ package main
+		return nil, nil, fmt.Errorf("Failed to parse yaml file, %s", err)
 	}
 
 	if labelsOverride != nil {
-		o.Labels = labelsOverride
+		pro.Labels = labelsOverride
 	}
-	return &o, nil
+	pro.SkipPullRequest = skipPullRequest
+	return &o, &pro, nil
+}
+
+func validateOptions(o *Options) error {
+	if len(o.Prefixes) == 0 {
+		return fmt.Errorf("Must have at least one Prefix specified")
+	}
+	if len(o.IncludedConfigPaths) == 0 {
+		return fmt.Errorf("includedConfigPaths is mandatory")
+	}
+	if o.TargetVersion != latestVersion && o.TargetVersion != upstreamVersion &&
+		o.TargetVersion != upstreamStagingVersion && !tagRegexp.MatchString(o.TargetVersion) {
+		logrus.Warnf("Warning: targetVersion is not one of %v so it might not work properly.",
+			[]string{latestVersion, upstreamVersion, upstreamStagingVersion, tagVersion})
+	}
+	if o.TargetVersion == upstreamVersion {
+		for _, prefix := range o.Prefixes {
+			if prefix.RefConfigFile == "" {
+				return fmt.Errorf("targetVersion can't be %q without refConfigFile for each prefix. %q is missing one", upstreamVersion, prefix.Name)
+			}
+		}
+	}
+	if o.TargetVersion == upstreamStagingVersion {
+		for _, prefix := range o.Prefixes {
+			if prefix.StagingRefConfigFile == "" {
+				return fmt.Errorf("targetVersion can't be %q without stagingRefConfigFile for each prefix. %q is missing one", upstreamStagingVersion, prefix.Name)
+			}
+		}
+	}
+	if (o.TargetVersion == upstreamVersion || o.TargetVersion == upstreamStagingVersion) && o.UpstreamURLBase == "" {
+		o.UpstreamURLBase = defaultUpstreamURLBase
+		logrus.Warnf("targetVersion can't be 'upstream' or 'upstreamStaging` without upstreamURLBase set. Default upstreamURLBase is %q", defaultUpstreamURLBase)
+	}
+
+	return nil
+}
+
+// UpdateReferences update the references of prow-images and/or boskos-images and/or testimages
+// in the files in any of "subfolders" of the includeConfigPaths but not in excludeConfigPaths
+// if the file is a yaml file (*.yaml) or extraFiles[file]=true
+func UpdateReferences(o *Options) (map[string]string, error) {
+	logrus.Info("Bumping image references...")
+	var allPrefixes []string
+	for _, prefix := range o.Prefixes {
+		allPrefixes = append(allPrefixes, prefix.Prefix)
+	}
+	filterRegexp := regexp.MustCompile(strings.Join(allPrefixes, "|"))
+	imageBumperCli := imagebumper.NewClient()
+	return updateReferences(imageBumperCli, filterRegexp, o)
+}
+
+type imageBumper interface {
+	FindLatestTag(imageHost, imageName, currentTag string) (string, error)
+	UpdateFile(tagPicker func(imageHost, imageName, currentTag string) (string, error), path string, imageFilter *regexp.Regexp) error
+	GetReplacements() map[string]string
+	AddToCache(image, newTag string)
+	TagExists(imageHost, imageName, currentTag string) (bool, error)
+}
+
+func updateReferences(imageBumperCli imageBumper, filterRegexp *regexp.Regexp, o *Options) (map[string]string, error) {
+	var tagPicker func(string, string, string) (string, error)
+	var err error
```

Comment:
not necessary to declare an error here, I'd move to just before the `upstreamImageVersionResolver()` call

Reply:
done